hoomd/mpcd/RejectionVirtualParticleFillerGPU.h

```diff
@@ -0,0 +1,171 @@
+                                          Scalar density,
+                                          unsigned int type,
+                                          std::shared_ptr<::Variant> T,
+                                          unsigned int seed,
+                                          std::shared_ptr<const Geometry> geom)
+        : mpcd::RejectionVirtualParticleFiller<Geometry>(sysdata, density, type, T, seed, geom),
+        m_track_bounded_particles(this->m_exec_conf), m_compact_idxs(this->m_exec_conf),
+        m_temp_storage(this->m_exec_conf)
+        {
+        m_tuner1.reset(new Autotuner(32, 1024, 32, 5, 100000, "mpcd_rejection_filler_draw_particles" + Geometry::getName(), this->m_exec_conf));
+        m_tuner2.reset(new Autotuner(32, 1024, 32, 5, 100000, "mpcd_rejection_filler_tag_particles" + Geometry::getName(), this->m_exec_conf));
+        }
+
+        //! Set autotuner parameters
+        /*!
+         * \param enable Enable/disable autotuning
+         * \param period period (approximate) in time steps when returning occurs
+         */
+        virtual void setAutotunerParams(bool enable, unsigned int period)
+            {
+            mpcd::RejectionVirtualParticleFiller<Geometry>::setAutotunerParams(enable, period);
+
+            m_tuner1->setEnabled(enable); m_tuner1->setPeriod(period);
+            m_tuner2->setEnabled(enable); m_tuner2->setPeriod(period);
+            }
+
+    protected:
+        //! Fill the volume outside the confinement
+        void fill(unsigned int timestep);
+        GPUArray<bool> m_track_bounded_particles; // Track if the particles are in/out of bounds for geometry
+        GPUArray<unsigned int> m_compact_idxs; // Indices for particles out of bound for geometry
+        GPUArray<unsigned int> m_temp_storage; // Temporary storage buffer for extracting compact indices array
+
+    private:
+        std::unique_ptr<::Autotuner> m_tuner1;   //!< Autotuner for drawing particles
+        std::unique_ptr<::Autotuner> m_tuner2;   //!< Autotuner for particle tagging
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFillerGPU<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = this->m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int N_virt_max = round(this->m_density*box.getVolume());
+
+    // Step 1: Create temporary arrays
+    if (N_virt_max > this->m_tmp_pos.getNumElements())
+        {
+        GPUArray<Scalar4> tmp_pos(N_virt_max, this->m_exec_conf);
+        GPUArray<Scalar4> tmp_vel(N_virt_max, this->m_exec_conf);
+        GPUArray<bool> track_bounded_particles(N_virt_max, this->m_exec_conf);
+        GPUArray<unsigned int> compact_idxs(N_virt_max, this->m_exec_conf);
+        GPUArray<unsigned int> temp_storage(N_virt_max, this->m_exec_conf);
```

Comment:
HOOMD has some amount of temporary memory available, which I usually use for these rather than a separate variable. The call pattern looks like this:
https://github.com/glotzerlab/hoomd-blue/blob/661faba150fd2d2a70a121a0408337cdbc21b666/hoomd/mpcd/ParticleData.cc#L1125-L1144

Reply:
Note that the size of this temporary storage is determined by CUB and not the number of particles, so you don't want to do the alloc here in any case.